sbol_utilities/default_components.py

```diff
@@ -0,0 +1,148 @@
+from typing import List, Tuple
+
+import sbol3
+import tyto
+
+def promoter(name: str, description: str, sequence: str)-> Tuple[sbol3.Component, sbol3.Sequence]:
+    """
+    Creates a promoter and a sequence Components with a given name, description, and sequence.
+    """
+    promoter_seq = sbol3.Sequence(f'{name}_seq')
+    promoter_seq.elements= sequence
+    promoter_seq.encoding = 'https://identifiers.org/edam:format_1207'
+
+    promoter = sbol3.Component(name, sbol3.SBO_DNA)
+    promoter.name = name
+    promoter.description = description
+    promoter.roles. append(sbol3.SO_PROMOTER)
+    promoter.sequences.append(promoter_seq)
+    return tuple([promoter, promoter_seq])
+
+def rbs(name: str, description: str, sequence: str)-> Tuple[sbol3.Component, sbol3.Sequence]:
+    """
+    Creates a Ribosome Binding Site (RBS) and a sequence Components with a given name, description, and sequence.
+    """
+    rbs_seq = sbol3.Sequence(f'{name}_seq')
```

Comment:
There's a lot of repeated structure in these functions. The common structure should be factored out into a shared function (`dna_component_with_sequence`?), such that each of these functions is a single line of code plus documentation. That will also make it easier to maintain and expand the collections.

Reply:
Yes they have the same pattern, the only difference is the role, that can be passed. How can we pass the argument ? It could be the URI 
```
ptet, ptet_seq = dna_component_with_sequence('sbol3.SO_PROMOTER', 'L3S2P21_UPA20_pTetR_RiboJ', 'BASIC - Potvin pTet (based on B-P27-pTetR-F1)','GTCCCTCG')
```
or as text that has a dictionary from  strings to URIs like {'promoter' : sbol3.SO_PROMOTER}. 
```
ptet, ptet_seq = dna_component_with_sequence('promoter', 'L3S2P21_UPA20_pTetR_RiboJ', 'BASIC - Potvin pTet (based on B-P27-pTetR-F1)','GTCCCTCG')
```
How does that affect the code readability. 
```
ptet, ptet_seq = promoter('L3S2P21_UPA20_pTetR_RiboJ', 'BASIC - Potvin pTet (based on B-P27-pTetR-F1)','GTCCCTCG')
```
I think the actual implementation is the most readable of the options but I never thought on maintainment nor its relation to collections. Can you explain me better those aspects please?

Reply:
I think that the current interface can stay the same. I am suggesting that the function `promoter` be implemented by a call to `dna_component_with_sequence`.  That way, `promoter`, `rbs`, `terminator`, etc do not contain any repeated code.

Reply:
Ah, now I understand, to make the repeated pattern in a function that then gets used by all the others. I will implement that.

Reply:
OK, here is a draft of that implementation, with using `dna_component_with_sequence` 

```
    def dna_component_with_sequence(identity: str, sequence: str, *, description: str)-> Tuple[sbol3.Component, sbol3.Sequence]:
    """
    Creates a Component and its Sequence.

    :param name: The name and identity of the Component. The identity of Sequence is also name with the suffix '_seq'.
    :param sequence: The DNA sequence of the Component encoded in IUPAC.
    :param description: The description of the Component.
    :return: A tuple of Component and Sequence.
    """
    comp_seq = sbol3.Sequence(f'{identity}_seq')
    comp_seq.elements= sequence
    comp_seq.encoding = 'sbol3.IUPAC_DNA_ENCODING'

    dna_comp = sbol3.Component(identity, sbol3.SBO_DNA)
    dna_comp.name = identity
    dna_comp.description = description
    dna_comp.sequences.append(comp_seq)
    return tuple([dna_comp, comp_seq])
```

Then to build a promoter the function should be like this:
```
    def promoter(identity: str, sequence: str, *, description: str)-> Tuple[sbol3.Component, sbol3.Sequence]:
    """
    Creates a Promoter Component and its Sequence.

    :param name: The name and identity of the Component. The identity of Sequence is also name with the suffix '_seq'.
    :param sequence: The DNA sequence of the Component encoded in IUPAC.
    :param description: The description of the Component.
    :return: A tuple of Component and Sequence.
    """
    promoter, promoter_seq = dna_component_with_sequence(identity, sequence, description=description)
    promoter.roles.append(sbol3.SO_PROMOTER)
    return tuple([promoter, promoter_seq])
```

Reply:
Yes, that's along the lines I'm thinking, plus the keyword arguments.

Reply:
Ok here is the implementation using kwargs and @tcmitchell constructor suggestion:
```
def dna_component_with_sequence(identity: str, sequence: str, **kwargs)-> Tuple[sbol3.Component, sbol3.Sequence]:
    """
    Creates a Component and its Sequence.

    :param identity: The identity of the Component. The identity of Sequence is also identity with the suffix '_seq'.
    :param sequence: The DNA sequence of the Component encoded in IUPAC.
    :param **kwargs: Any other Component attribute.
    :return: A tuple of Component and Sequence.
    """
    comp_seq = sbol3.Sequence(f'{identity}_seq',
                                elements=sequence,
                                encoding='sbol3.IUPAC_DNA_ENCODING')

    dna_comp = sbol3.Component(identity, sbol3.SBO_DNA,
                                sequences =[comp_seq],
                                **kwargs)
    return tuple([dna_comp, comp_seq])

def promoter(identity: str, sequence: str, **kwargs)-> Tuple[sbol3.Component, sbol3.Sequence]:
    """
    Creates a Promoter Component and its Sequence.

    :param identity: The identity of the Component. The identity of Sequence is also identity with the suffix '_seq'.
    :param sequence: The DNA sequence of the Component encoded in IUPAC.
    :param **kwargs: Any other Component attribute.
    :return: A tuple of Component and Sequence.
    """
    promoter, promoter_seq = dna_component_with_sequence(identity, sequence, **kwargs)
    promoter.roles.append(sbol3.SO_PROMOTER)
    return tuple([promoter, promoter_seq])
```

how should I indicate to the user how to use the kwarg? I used :param **kwargs: Any other Component attribute.

Reply:
That's looking pretty good to me.
I don't know the most pythonic way of documenting kwargs; ask the internet?